### PR TITLE
feat(cdn): artifact support

### PR DESCRIPTION
**Pull request by bnjjj**
1. Description
1. Related issues
1. About tests
1. Mentions

@ovh/cds

### CODE REVIEW

File: engine/cdn/cdn.go [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+	if err := s.initDefaultDrivers(ctx); err != nil {
+		return sdk.WrapError(err, "cannot init drivers")
+	}
+
+	//Init the http server
+	s.initRouter(ctx)
+	server := &http.Server{
+		Addr:           fmt.Sprintf("%s:%d", s.Cfg.HTTP.Addr, s.Cfg.HTTP.Port),
+		Handler:        s.Router.Mux,
+		ReadTimeout:    10 * time.Minute,
```
Comment (fsamin) [inline engine/cdn/cdn.go]: For a content de livery service, it is weird to have a readTimeout and writeTimeout

File: engine/cdn/mirrors.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+	"github.com/ovh/cds/engine/cdn/objectstore"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/log"
+)
+
+func (s *Service) mirroring(object objectstore.Object, body io.Closer, reader io.Reader) {
+	defer body.Close()
+
+	for _, mirror := range s.MirrorDrivers {
```
Comment (fsamin) [inline engine/cdn/mirrors.go]: see above

File: engine/cdn/objectstore/objectstore.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+	Status() sdk.MonitoringStatusLine
+	Store(o Object, data io.ReadCloser) (string, error)
+	ServeStaticFiles(o Object, entrypoint string, data io.ReadCloser) (string, error)
+	Fetch(o Object) (io.ReadCloser, error)
+	Delete(o Object) error
+	TemporaryURLSupported() bool
+}
+
+// DriverWithRedirect has to be implemented if your storage backend supports temp url
+type DriverWithRedirect interface {
```
Comment (fsamin) [inline engine/cdn/objectstore/objectstore.go]: I think the temp URL are useless with this new service

File: engine/cdn/objectstore/types.go [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package objectstore
+
+//Object is the interface for stuff needed to be stored in object store
+type Object interface {
```
Comment (fsamin) [inline engine/cdn/objectstore/types.go]: is this still useful ?

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11209.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report it#11209.0 ✘  
* w-it
  * w-it ✘

File: engine/cdn/objectstore/types.go (hunk 4 above)
Comment (bnjjj) [reply engine/cdn/objectstore/types.go]: Yes I use it for each type of object to store. It's very handy

File: engine/cdn/objectstore/objectstore.go (hunk 3 above)
Comment (bnjjj) [reply engine/cdn/objectstore/objectstore.go]: To delete.

File: engine/api/authentication/cdn/cdn.go [hunk 5]
```diff
@@ -0,0 +1,28 @@
+package worker
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply engine/api/authentication/cdn/cdn.go]: oups

File: engine/api/workflow_queue_storage.go [hunk 6]
```diff
@@ -121,153 +124,12 @@ func (api *API) postWorkflowJobStaticFilesHandler() service.Handler {
...
-		nodeRun.Artifacts = append(nodeRun.Artifacts, art)
-		if err := workflow.InsertArtifact(api.mustDB(), &art); err != nil {
-			_ = storageDriver.Delete(&art)
-			return sdk.WrapError(err, "Cannot update workflow node run")
-		}
-		return nil
-	}
-}
-
 func (api *API) postWorkflowJobArtifactWithTempURLCallbackHandler() service.Handler {
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply engine/api/workflow_queue_storage.go]: Yes I have to rename it. TempURL is implicit with cdn service

File: engine/cdn/cdn_handlers.go [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+		var file io.ReadCloser
+		switch cdnRequest.Type {
+		case sdk.CDNArtifactType:
+			if cdnRequest.Artifact == nil {
+				return fmt.Errorf("cannot download artifact, need artifact description in cdn request token")
+			}
+			w.Header().Add("Content-Disposition", fmt.Sprintf("attachment; filename=\"%s\"", cdnRequest.Artifact.Name))
+
+			var err error
+			file, err = s.downloadArtifact(r, *cdnRequest)
```
[the remark this answers is not in the document]
Comment (bnjjj) [reply engine/cdn/cdn_handlers.go]: file is a stream

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11215.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report it#11215.0 ✘  
* w-it
  * w-it ✘

File: engine/api/workflow_queue_storage.go [hunk 8]
```diff
@@ -280,61 +141,39 @@ func (api *API) postWorkflowJobArtifactWithTempURLCallbackHandler() service.Hand
...
+			// TODO: call to cdn to delete artifact
 			return sdk.WrapError(err, "Cannot update workflow node run")
 		}
 
 		return nil
 	}
 }
 
-func (api *API) postWorkflowJobArtifacWithTempURLHandler() service.Handler {
+func (api *API) postWorkflowJobArtifactHandler() service.Handler {
```
Comment (sguiheux) [inline engine/api/workflow_queue_storage.go]: Worker must call CDN Directly

File: engine/cdn/cdn_handlers.go [hunk 9]
```diff
@@ -0,0 +1,108 @@
+package cdn
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+
+	"github.com/gorilla/mux"
+	cdnauth "github.com/ovh/cds/engine/api/authentication/cdn"
```
Comment (sguiheux) [inline engine/cdn/cdn_handlers.go]: need empty line

File: engine/cdn/mirrors.go [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+	"github.com/ovh/cds/engine/cdn/objectstore"
+	"github.com/ovh/cds/sdk"
+	"github.com/ovh/cds/sdk/log"
+)
+
+func (s *Service) mirroring(object objectstore.Object, reader io.Reader) {
+	for _, mirror := range s.MirrorDrivers {
+		var buf bytes.Buffer
+		tee := io.TeeReader(reader, &buf)
+		_, err := mirror.Store(context.Background(), object, ioutil.NopCloser(tee))
```
Comment (sguiheux) [inline engine/cdn/mirrors.go]: Do . we need something to re enqueue ?

File: engine/cdn/objectstore/swift.go [hunk 11]
```diff
@@ -0,0 +1,262 @@
...
+		return "", sdk.WrapError(err, "Unable to close data buffer")
+	}
+
+	return s.StorageUrl + "/" + container, nil
+
+}
+
+// Store stores in swift
+func (s *SwiftStore) Store(ctx context.Context, o Object, data io.ReadCloser) (string, error) {
+	container := s.containerPrefix + o.GetPath()
```
Comment (sguiheux) [inline engine/cdn/objectstore/swift.go]: Need long retry

File: engine/cdn/objectstore/swift.go [hunk 12]
```diff
@@ -0,0 +1,262 @@
...
+	if err := data.Close(); err != nil {
+		return "", sdk.WrapError(err, "Unable to close data buffer")
+	}
+
+	return container + "/" + object, nil
+}
+
+// Fetch an object from swift
+func (s *SwiftStore) Fetch(_ context.Context, o Object) (io.ReadCloser, error) {
+	container := s.containerPrefix + o.GetPath()
```
Comment (sguiheux) [inline engine/cdn/objectstore/swift.go]: Need retry ?

File: engine/cdn/artifacts.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+	}
+
+	id := storageDriver.GetProjectIntegration().ID
+	if id > 0 {
+		art.ProjectIntegrationID = &id
+	}
+
+	var buf bytes.Buffer
+	tee := io.TeeReader(body, &buf)
+	objectPath, err := storageDriver.Store(ctx, art, ioutil.NopCloser(tee))
```
Comment (sguiheux) [inline engine/cdn/artifacts.go]: Retry ?

File: engine/cdn/artifacts.go [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+
+	"github.com/ovh/cds/sdk"
+)
+
+func (s *Service) storeArtifact(ctx context.Context, body io.ReadCloser, cdnRequest sdk.CDNRequest) (*sdk.WorkflowNodeRunArtifact, error) {
+	art := cdnRequest.Artifact
+	if _, err := art.IsValid(); err != nil {
+		return nil, sdk.WrapError(err, "artifact is not valid")
+	}
+	storageDriver, err := s.getDriver(cdnRequest.ProjectKey, cdnRequest.IntegrationName)
```
Comment (sguiheux) [inline engine/cdn/artifacts.go]: Why not push it in first layer cache , and then upload in storage + mirror after ?   to be fast

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11282.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report it#11282.0 ✘  
* w-it
  * w-it ✘

**Comment by ovh-cds** [comment]
CDS Report it#11295.0 ✘  
* w-it
  * w-it ✘

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11297.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11298.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report it#11298.0 ✘  
* w-it
  * w-it ✘

**Comment by ovh-cds** [comment]
CDS Report build-engine-cli-tools#11300.0 ✘  
* Build minimal
  * cdsctl linux ✔ 
  * check SQL Files number ✔ 
  * Engine Build linux amd64 ✔ 
  * tools/debpacker ✔ 
  * tools/os-ansible-inventory ✘  
  * Worker linux amd64 / 386 ✔

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11300.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report build-ui#11302.0 ✘  
* Compile
  * Build UI ✔ 
  * Lint Translate ✔ 
  * TU All -   
  * TU model ✘  
  * TU shared ✔ 
  * TU store ✔ 
  * TU views ✔

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11302.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report ut-engine#11304.0 ✘  
* Stage 1
  * Engine Test with DB ✘

**Comment by ovh-cds** [comment]
CDS Report it#11304.0 ✘  
* w-it
  * w-it ✘

**Pull request closed without merging**